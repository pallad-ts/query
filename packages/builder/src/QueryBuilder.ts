import {Builder} from "@pallad/builder";
import {SchemaValidation, Validation, ValidatorError, ViolationsList} from "alpha-validator";
import * as yup from 'yup';
import {createSortableFieldSchema} from "./createSortableFieldSchema";
import {byYup} from "alpha-validator-bridge-yup";
import {QueryConfig} from "./QueryConfig";
import {List, Maybe} from "monet";

function concatSchema(schema: yup.ObjectSchema<any>, anotherSchema: yup.ObjectSchema<any>) {
	return schema.concat(anotherSchema);
}

export class QueryBuilder<T extends QueryConfig<any, any>> extends Builder {
	private filtersSchemaObject = yup.object()
		.default({})

	private validation?: (data: unknown) => Promise<Validation<ViolationsList, T['query']>>;

	constructor(readonly config: T) {
		super();
	}

	private getPaginationSchema(): Maybe<yup.ObjectSchema<any>> {
		return Maybe.fromUndefined(
			this.config.getPagination()
		)
			.map(x => {
				if (x.type === 'byCursor') {
					return yup.object().shape({
						limit: this.getLimitSchema(x),
						after: yup.string(),
						before: yup.string()
					});
				}
				return yup.object().shape({
					limit: this.getLimitSchema(x),
					offset: yup.number()
						.integer()
						.min(0)
						.default(0)
				});
			});
	}

	private getLimitSchema(paginationConfig: QueryConfig.Pagination) {
		return yup.number()
			.min(1)
			.max(paginationConfig.maxLimit)
			.integer()
			.default(paginationConfig.defaultLimit);
	}

	filtersSchema(schema: yup.ObjectSchema<any>) {
		this.validation = undefined;
		this.filtersSchemaObject = schema.clone().default({});
		return this;
	}

	private getSortingSchema(): Maybe<yup.ObjectSchema<any>> {
		return Maybe.fromUndefined(
			this.config.getSorting()
		).map(x => {
			const sortByType = createSortableFieldSchema(x.sortableFields.slice());
			if (x.type === 'single') {
				return yup.object().shape({
					sortBy: sortByType.default(x.defaultSorting)
				});
			}

			return yup.object().shape({
				sortBy: yup.array()
					.of(sortByType)
					.min(1)
					.default(x.defaultSorting)
			});
		});
	}

	build(input: unknown): Promise<Validation<ViolationsList, T['query']>> {
		return this.getValidationFunction()(input);
	}

	async buildOrFail(input: unknown, errorMessage?: string) {
		const result = await this.build(input);
		if (result.isFail()) {
			throw new ValidatorError(result.fail(), errorMessage || 'Invalid query');
		}
		return result.success();
	}

	private getValidationFunction() {
		if (!this.validation) {
			let schema = yup.object()
				.shape({
					filters: this.filtersSchemaObject
				})
				.noUnknown();

			schema = this.getPaginationSchema().foldLeft(schema)(concatSchema);
			schema = this.getSortingSchema().foldLeft(schema)(concatSchema)

			this.validation = SchemaValidation.toValidationFunction<unknown, T['query'], undefined>(
				'query builder',
				byYup<unknown, T['query']>(schema)
			) as (data: unknown) => Promise<Validation<ViolationsList, T['query']>>;
		}
		return this.validation;
	}
}


export namespace QueryBuilder {

}
