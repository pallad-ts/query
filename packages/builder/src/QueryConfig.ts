import {PaginableByCursor, PaginableByOffset, Query, SortableFieldDefinition, SortableMulti, SortableSingle} from "@pallad/query";
import {Builder} from "@pallad/builder";
import {QueryBuilder} from "./QueryBuilder";
import {ERRORS} from "./errors";
import {ResultBuilder} from "./ResultBuilder";

function composeDefaultPaginationOptions<T extends Partial<Omit<QueryConfig.Pagination, 'type'>>>(options?: T) {
	return {
		...QueryConfig.defaultPaginationOptions,
		...(options || {})
	};
}

export class QueryConfig<TFilters, TQueryProperties, TResultMeta = unknown> extends Builder {
	private sorting?: QueryConfig.Sorting<any>;
	private pagination?: QueryConfig.Pagination;

	readonly query!: Query<TFilters> & TQueryProperties;
	readonly resultMeta!: TResultMeta;

	static defaultPaginationOptions: Omit<QueryConfig.Pagination, 'type'> = {
		defaultLimit: 50,
		maxLimit: 1000
	};

	cursorPagination(
		options?: Partial<Omit<QueryConfig.Pagination.ByCursor, 'type'>>
	): QueryConfig<TFilters, TQueryProperties & PaginableByCursor, TResultMeta & PaginableByCursor.ResultMeta> {
		this.pagination = {
			type: 'byCursor',
			...composeDefaultPaginationOptions(options)
		};

		return this as any;
	}

	offsetPagination(
		options?: Partial<Omit<QueryConfig.Pagination.ByOffset, 'type'>>
	): QueryConfig<TFilters, TQueryProperties & PaginableByOffset, TResultMeta & PaginableByOffset.ResultMeta> {
		this.pagination = {
			type: 'byOffset',
			...composeDefaultPaginationOptions(options)
		};
		return this as any;
	}

	getPagination() {
		return this.pagination;
	}

	singleSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: SortableFieldDefinition<TSortableField>
	) {
		this.sorting = {
			type: 'single',
			sortableFields,
			defaultSorting
		};
		return this as unknown as QueryConfig<TFilters,
			TQueryProperties & SortableSingle<TSortableField>,
			TResultMeta & SortableSingle.ResultMeta<TSortableField>>;
	}

	multiSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: Array<SortableFieldDefinition<TSortableField>>
	) {
		this.sorting = {
			type: 'multi',
			sortableFields,
			defaultSorting
		};
		return this as unknown as QueryConfig<TFilters,
			TQueryProperties & SortableMulti<TSortableField>,
			TResultMeta & SortableMulti.ResultMeta<TSortableField>>;
	}

	getSorting() {
		return this.sorting;
	}

	filters<T>() {
		return this as unknown as QueryConfig<T, TQueryProperties>;
	}

	getQueryBuilder(): QueryBuilder<this> {
		this.validate();
		return new QueryBuilder(this);
	}

	validate() {
		if (this.pagination && this.pagination.type === 'byCursor') {
			if (!this.sorting) {
				throw ERRORS.MISSING_SINGLE_SORTING_FOR_CURSOR_PAGINATION();
			}

			if (this.sorting.type === 'multi') {
				throw ERRORS.MULTI_SORTING_FOR_ALLOWED_FOR_CURSOR_PAGINATION();
			}
		}
	}

	getResultBuilder(): ResultBuilder<this> {
		this.validate();
		return new ResultBuilder(this);
	}
}

export namespace QueryConfig {
	export type Pagination = Pagination.ByCursor | Pagination.ByOffset;
	export namespace Pagination {
		export interface ByCursor {
			type: 'byCursor';
			defaultLimit: number;
			maxLimit: number;
		}

		export interface ByOffset {
			type: 'byOffset';
			defaultLimit: number;
			maxLimit: number;
		}
	}

	export type Sorting<TSortableField extends string> = Sorting.Single<TSortableField> | Sorting.Multi<TSortableField>;
	export namespace Sorting {
		export interface Single<TSortableField extends string> {
			type: 'single';
			sortableFields: TSortableField[];
			defaultSorting: SortableFieldDefinition<TSortableField>
		}

		export interface Multi<TSortableField extends string> {
			type: 'multi';
			sortableFields: readonly TSortableField[];
			defaultSorting: Array<SortableFieldDefinition<TSortableField>>;
		}
	}
}
