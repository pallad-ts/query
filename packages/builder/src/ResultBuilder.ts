import {PaginableByCursor, PaginableByOffset, Query, Result, ResultMeta, SortableFieldDefinition, SortableMulti, SortableSingle} from "@pallad/query";
import {QueryConfig} from "./QueryConfig";
import {Builder} from "@pallad/builder";
import {Maybe} from "monet";

export class ResultBuilder<T extends QueryConfig<any, any>, TDefaultEntity> extends Builder {
	constructor(readonly config: T) {
		super();
	}

	build<TEntity = TDefaultEntity>(options: ResultBuilder.Options<TEntity, T['query']>) {
		let meta: any = {};
		this.composeMetaForPagination(options).foldLeft(meta)(Object.assign)
		this.composeMetaForSorting(options).foldLeft(meta)(Object.assign);

		const hasMeta = Object.keys(meta).length > 0;

		return {
			results: options.results,
			...(hasMeta ? {meta} : {})
		} as Result<TEntity> & (T['resultMeta'] extends {} ? ResultMeta<T['resultMeta']> : unknown);
	}

	private composeMetaForPagination(options: ResultBuilder.Options<unknown, any>): Maybe<PaginableByCursor.ResultMeta | PaginableByOffset.ResultMeta> {
		return Maybe.fromUndefined(this.config.getPagination())
			.map(settings => {
				if (settings.type === 'byOffset') {
					const query = options.query as Required<PaginableByOffset>;
					return {
						limit: query.limit,
						offset: query.offset
					} as PaginableByOffset.ResultMeta;
				}

				const query = options.query as Required<PaginableByCursor>;
				return {
					limit: query.limit,
					nextPage: options.nextPageCursor,
					previousPage: options.previousPageCursor
				} as PaginableByCursor.ResultMeta;
			});
	}

	private composeMetaForSorting({query}: ResultBuilder.Options<unknown, any>): Maybe<SortableSingle.ResultMeta<any> | SortableMulti.ResultMeta<any>> {
		return Maybe.fromUndefined(this.config.getSorting())
			.map(() => {
				return {sortBy: query.sortBy}
			});
	}
}

export namespace ResultBuilder {
	export interface Options<TEntity, TQuery extends Query<any>> {
		results: TEntity[];
		query: TQuery;
		nextPageCursor?: string;
		previousPageCursor?: string;
	}
}

type Test<T> = { foo: 'bar' } & (T extends {} ? { bar: T } : unknown);

const o: Test<{ zee: 'zee' }> = {
	foo: 'bar',
	bar: {zee: 'zee'}
}

const o2: Test<unknown> = {
	foo: 'bar'
}
