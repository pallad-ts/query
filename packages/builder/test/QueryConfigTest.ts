import {QueryConfig} from "@src/QueryConfig";
import {assert, IsExact} from "conditional-type-checks";
import {PaginableByCursor, PaginableByOffset, Query, SortableMulti, SortableSingle} from "@pallad/query";
import {QueryBuilder} from "@src/QueryBuilder";
import {ResultBuilder} from "@src/ResultBuilder";
import {Filters} from "./fixtures";

describe('QueryConfig', () => {
	describe('pagination', () => {
		it('by default not set', () => {
			const config = new QueryConfig();
			expect(config.getPagination())
				.toBeUndefined();
		});

		describe('by cursor', () => {
			it('types', () => {
				const config = new QueryConfig().cursorPagination();
				type ExpectedQuery = Query<unknown> & PaginableByCursor;
				type ExpectedResultMeta = PaginableByCursor.ResultMeta;

				assert<IsExact<typeof config['query'], ExpectedQuery>>(true);
				assert<IsExact<typeof config['resultMeta'], ExpectedResultMeta>>(true);
			})

			it('applies default config if not set explicitly', () => {
				const config = new QueryConfig().cursorPagination();

				expect(config.getPagination())
					.toEqual({
						type: 'byCursor',
						defaultLimit: 50,
						maxLimit: 1000
					})
			});

			it('merges with default settings', () => {
				const config = new QueryConfig().cursorPagination({defaultLimit: 10});

				expect(config.getPagination())
					.toEqual({
						type: 'byCursor',
						defaultLimit: 10,
						maxLimit: 1000
					});
			});
		});

		describe('by offset', () => {
			it('types', () => {
				const config = new QueryConfig().offsetPagination();
				type ExpectedQuery = Query<unknown> & PaginableByOffset;
				type ExpectedResultMeta = PaginableByOffset.ResultMeta;

				assert<IsExact<typeof config['query'], ExpectedQuery>>(true);
				assert<IsExact<typeof config['resultMeta'], ExpectedResultMeta>>(true);
			})

			it('applies default config if not set explicitly', () => {
				const config = new QueryConfig().offsetPagination();

				expect(config.getPagination())
					.toEqual({
						type: 'byOffset',
						defaultLimit: 50,
						maxLimit: 1000
					})
			});

			it('merges with default settings', () => {
				const config = new QueryConfig().offsetPagination({defaultLimit: 10});

				expect(config.getPagination())
					.toEqual({
						type: 'byOffset',
						defaultLimit: 10,
						maxLimit: 1000
					});
			});
		});
	});

	describe('sorting', () => {
		it('by default not set', () => {
			const config = new QueryConfig();
			expect(config.getSorting())
				.toBeUndefined();
		});

		describe('single', () => {
			const config = new QueryConfig()
				.singleSorting(['foo', 'bar'], {field: 'foo', direction: 'ASC'});

			it('types', () => {
				type ExpectedQuery = Query<unknown> & SortableSingle<'foo' | 'bar'>;
				type ExpectedResultMeta = SortableSingle.ResultMeta<'foo' | 'bar'>;

				assert<IsExact<typeof config['query'], ExpectedQuery>>(true);
				assert<IsExact<typeof config['resultMeta'], ExpectedResultMeta>>(true);
			});

			it('getting config', () => {
				expect(config.getSorting())
					.toEqual({
						type: 'single',
						sortableFields: ['foo', 'bar'],
						defaultSorting: {field: 'foo', direction: 'ASC'}
					});
			});
		});

		describe('multi', () => {
			const config = new QueryConfig()
				.multiSorting(['foo', 'bar'], [
					{field: 'foo', direction: 'ASC'},
					{field: 'bar', direction: 'DESC'}
				]);

			it('types', () => {
				type ExpectedQuery = Query<unknown> & SortableMulti<'foo' | 'bar'>;
				type ExpectedResultMeta = SortableMulti.ResultMeta<'foo' | 'bar'>;

				assert<IsExact<typeof config['query'], ExpectedQuery>>(true);
				assert<IsExact<typeof config['resultMeta'], ExpectedResultMeta>>(true);
			});

			it('getting config', () => {
				expect(config.getSorting())
					.toEqual({
						type: 'multi',
						sortableFields: ['foo', 'bar'],
						defaultSorting: [
							{field: 'foo', direction: 'ASC'},
							{field: 'bar', direction: 'DESC'}
						]
					});
			});
		});
	});

	describe('filters', () => {
		it('just changes type of filters', () => {
			const config = new QueryConfig()
				.filters<Filters>();

			type Expected = Query<Filters>;
			assert<IsExact<typeof config['query'], Expected>>(true);
		});
	});

	describe('query builder', () => {
		it('required config to be valid before returning query builder', () => {
			expect(() => {
				new QueryConfig().cursorPagination().getQueryBuilder();
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('types', () => {
			const builder = new QueryConfig()
				.singleSorting(['foo', 'bar'], {field: 'bar', direction: 'ASC'})
				.filters<Filters>()
				.getQueryBuilder();

			type Expected = QueryBuilder<QueryConfig<Filters, SortableSingle<'foo' | 'bar'>>>;
			assert<IsExact<typeof builder, Expected>>(true);
		});
	});

	describe('result builder', () => {
		it('required config to be valid before returning result builder', () => {
			expect(() => {
				new QueryConfig().cursorPagination().getResultBuilder();
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('types', () => {
			const builder = new QueryConfig()
				.singleSorting(['foo', 'bar'], {field: 'bar', direction: 'ASC'})
				.cursorPagination()
				.filters<Filters>()
				.getResultBuilder();

			type Expected = ResultBuilder<QueryConfig<Filters, SortableSingle<'foo' | 'bar'> & PaginableByCursor>>;
			assert<IsExact<typeof builder, Expected>>(true);
		});
	});

	describe('validation', () => {
		it('pagination by cursor cannot be used without single sorting setup', () => {
			expect(() => {
				new QueryConfig().cursorPagination().validate();
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('pagination by cursor cannot be used with multi sorting', () => {
			expect(() => {
				new QueryConfig()
					.cursorPagination()
					.multiSorting(['foo', 'bar'], [{field: 'foo', direction: 'ASC'}]).validate();
			})
				.toThrowErrorMatchingSnapshot();
		});
	})
});
