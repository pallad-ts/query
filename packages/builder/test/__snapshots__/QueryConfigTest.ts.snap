// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryConfig query builder required config to be valid before returning query builder 1`] = `"Pagination by cursor cannot be used without single field sorting"`;

exports[`QueryConfig result builder required config to be valid before returning result builder 1`] = `"Pagination by cursor cannot be used without single field sorting"`;

exports[`QueryConfig validation pagination by cursor cannot be used with multi sorting 1`] = `"Sorting by multiple fields is not allowed in pagination by cursor"`;

exports[`QueryConfig validation pagination by cursor cannot be used without single sorting setup 1`] = `"Pagination by cursor cannot be used without single field sorting"`;
